README.md
# Quick Item Swap

Quick Item Swap is a mod that enables you to quickly change between related items.  For instance, if you're currently
holding a transport belt, the "Next Item Group" keybind will pick up a Fast Transport Belt instead (presuming you 
have one) and the "Next Related Item" keybind will pick up a Splitter instead (again presuming you have one).  By
default, 

A wide variety of items and groups are supported, including transport belts, inserters, train vehicles,
 railway and rail signals, inserters, pipes, and modules.     

New in version 0.3.2 is the ability for item cycling to place ghosts of items you do not have, and the ability to 
whitelist or blacklist specific items. 

 
 Additionally, the following mods are supported.
 
  * [Batteries Not Included](https://mods.factorio.com/mods/dewiniaid/BatteriesNotIncluded)
  * [Electric Vehicles](https://mods.factorio.com/mod/ElectricVehicles3) (locomotive only)
  * [FARL](https://mods.factorio.com/mod/FARL) (locomotive only)
  * [Creative Mode](https://mods.factorio.com/mod/creative-mode-fix)
  * [Logistic Train Network](https://mods.factorio.com/mod/LogisticTrainNetwork)
  * [Train Supply Manager](https://mods.factorio.com/mod/train-pubsub)
  * [Smarter Trains](https://mods.factorio.com/mod/SmartTrains)
  * [Vehicle Wagon](https://mods.factorio.com/mod/Vehicle%20Wagon)
  * [Nixie Tubes](https://mods.factorio.com/mod/nixie-tubes) (Just justarandomgeek's version currently)
   
## Known Issues

* Items on the quickbar without filters may change locations if the mod selects them, and the quickbar in
  general may have some odd interactions.  These are harmless (but annoying) and have no good fix until
  Factorio 0.16
  
* Mousewheel bindings do not work consistently due to what appears to be a
  [bug in Factorio](https://forums.factorio.com/viewtopic.php?f=34&t=54327).
  
* Not tested in multiplayer.  Please report back!

## Unknown Issues
   
Found a bug?  Please visit the [Issues page](https://github.com/dewiniaid/QuickItemSwap/issues) to see if it has 
already been reported, and report it if not.

## Changelog

### 0.3.4 (2020-06-14)
* Fix crash for 'Unknown consuming type' when loading mod

### 0.3.3 (2020-01-31)
* **Update for Factorio 0.18**

### 0.3.2 (2019-03-14)
*a.k.a. version 0.3.14159*
* QuickItemSwap can now cycle between item ghosts if you lack the actual item.
  * By default, this excludes items that have not been researched.  You can globally disable this in Mod Settings, or you can whitelist specific items.
* You can now whitelist or blacklist items (default keybind: Alt+Y while holding an item)
  * When cycling between related items, blacklisted items will be ignored (unless they are in your inventory)
  * When cycling between related items, whitelisted items will be allowed (even if they are not yet researched)
  * Blacklisting and whitelisting only affects cases when a ghost would be chosen or if an item would be created (in cheat mode).  Items in your inventory are always considered.
  * You can reset your entire blacklist with `/qis-clear-blacklist`
* Added support for the train stops in [Train Supply Manager](https://mods.factorio.com/mod/train-pubsub)
* Fixed possible crash when cycling blueprints.

### 0.3.1 (2019-02-27)
* **Update for Factorio 0.17.**
* Removed all quickbar logic.
* Generate group information for belts/splitters/undergrounds procedurally.  This should automatically add support for modded belts, such as those from Bob's Logistics.

### 0.3.0 (2018-04-23)
* Add support for blueprints, blueprint books, deconstruction planners, and similar items from mods.
  *  Items of a particular type are treated as members of one group.  "Y" will cycle between different blueprint books, for instance.
  * Cycling groups will cycle between selecting a blueprint, a blueprint book, or a deconstruction planner.
  * Other items that have an item_number are also supported when cycling items, but do not support cycling groups.
* Added support for the Artillery Wagon
* Add support for [Electric Vehicles](https://mods.factorio.com/mod/ElectricVehicles3) (locomotives only)
* Add support for [FARL](https://mods.factorio.com/mod/FARL) (locomotives only)

### 0.2.4 (2017-12-23)
* Add support for the unofficial Creative Mode Fix.
* Add Buffer Chests to the Logistic Chests group, and reordered the group to match the inventory window.

### 0.2.3 (2017-12-13)
* **Update for Factorio 0.16**
* Temporarily add support for Batteries Not Included, since [a bug in the current Factorio experimental](https://forums.factorio.com/viewtopic.php?f=182&t=54567&p=321491) prevents that support from working in the opposite direction.
* API: Added more verbose error messaging if another mod submits an invalid patch.
* API: Swapped the order of arguments on `apply_patch` to make it less likely for someone to inadvertently forget to
  specify `source`.

### 0.2.2 (2017-12-11)
* Add support for items from Logistic Train Network, Smarter Trains, Vehicle Wagon, and Nixie Tubes.
* Add speculative support for artillery wagons.

### 0.2.1 (2017-12-05)
* Fix default keybindings for Previous Item and Previous Group.
* Console announcements when a configuration change is detected are now only displayed in debug mode.  They also now 
  are prefixed with `[QuickItemSwap]` to indicate they are coming from this mod.

### 0.2.0 (2017-12-01)
* Made significant internal changes to allow mod integration
* Added support for items from Creative Mode
* Added support for creating items from thin air (rather than skipping over them) when your player is in cheat mode (i.e. as a result of the creative mode mode or `/c game.players[1].cheat_mode = true`).  Can be toggled in settings.  Has no effect when not in cheat mode.
* First pass of an API to allow other mods to add recipes.

### 0.1.1 (2017-11-30)
* Flip logic on modules to make more sense: each tier is now a group, rather than each module type being a group.
 
### 0.1.0 (2017-11-30)
* First public release.

## API for Mod Developers 

For discussion on this topic, find me on IRC or Discord or 
[comment on this issue](https://github.com/dewiniaid/QuickItemSwap/issues/3).

General recommended flow is:

 - Make your mod optionally depend on QuickItemSwap so that QuickItemSwap initializes first.
 - During your mod's `on_init`, `on_configuration_changed` and possibly `on_runtime_mod_settings_changed` events, call
   `apply_patch` with the appropriate information.
 - Listen for the custom events `on_qis_mappings_reset` and, if desired, `on_qis_mappings_patched` (event IDs are 
   published through `get_events`) and resubmit your patches as needed. 

Currently implemented:

#### `remote.call("QuickItemSwap", "debug", [new_setting])`

Returns (and possibly changes) the debug setting for QIS. Note that debug being on will produce an incessant amount
of screen spam.  Set to `true` to enable debugging, `false` to disable, or `nil` to just see what the current setting is.
 
#### `remote.call("QuickItemSwap", "dump_mappings")`

Writes the current mapping table to `script-output/quickitemswap-mappings.txt`

#### `remote.call("QuickItemSwap", "get_events")`

Returns a table of `string = number` containing the custom events used by QuickItemSwap.  Right now, the two events
are:
  
  - `on_qis_mappings_reset` - Triggered when QuickItemSwap or another mod resets the item mappings and rebuilds them
    from scratch.  This can happen on version updates, mod changes, etc.  When this fires, any patches provided by your
    mod will be lost and thus must be resubmitted.
  
    Event table fields:
    
    | Field | Explanation |
    | ----- | ----------- |
    | `why` | Reason for the reset.  May be `"init"` (QIS's `on_init()`), `"settings-changed"` (QIS's mod settings changing), `"configuration-changed"` (Mod versions or startup settings changing) or `"remote"` (a remote caller asked it to reset)  

  - `on_qis_mappings_patched` - Triggered in response to someone (possibly you) calling `apply_patch`.  Allows for mods to react to other mods patching the table if it makes sense to.   
  
    Event table fields:
    
    | Field | Explanation |
    | ----- | ----------- |
    | `patch` | Contents of the patch that was provided.
    | `source` | Source providing the patch.  This is dependant on the caller providing a meaningful `source` parameter.
    
#### `remote.call("QuickItemSwap", "get_mappings", [category, [group, [item]]])`

Returns information on currently defined mappings.  If all 3 parameters are nil, the entire table is returned.  
Otherwise, returns just a subset for a given category, group within a category, or item within a group.

#### `remote.call("QuickItemSwap", "refresh", [only_if_dirty])`

Refreshes the internal state that QIS maintains to properly cycle between things.  If the optional parameter is `true`,
the refresh is only performed if the data is marked as out of date.

This is generally unneccessary -- `apply_patch` marks the data as outdated and it will be rebuilt on the first access
that requires it when outdated.

If `player` is specified, results will be printed to that player; otherwise they are printed using `game.print`
 
Produces no output if everything is valid.

#### `remote.call("QuickItemSwap", "validate_mappings", [player])`

Prints to the console any items in the mapping table that don't have a corresponding item prototype.

If `player` is specified, results will be printed to that player; otherwise they are printed using `game.print`
 
Produces no output if everything is valid.

#### `remote.call("QuickItemSwap", "apply_patch", patch, source)`

Applies a patch to the mapping table.

`source` is the name of your mod, and is used to explains where the patch came from.  It appears in error messages and 
in the `on_qis_mappings_patched` event.  It is recommended you use the `info.json` name of your mod here for
consistency, though nothing enforces this.

`patch` consists of a table formatted similar to the mapping table.  (See `mappings/base.lua` for a reference.)  It will be merged into the mapping table using
the following rules:

- If a category does not exist, it will be created.  If it exists, the groups will be merged.  If the patch explicitly 
  sets the category to `false`, the category will be deleted.  If the patch omits the category or sets it to `nil`, it 
  will be left as-is.
  
- The above logic also applies to groups within a category (with items being merged) and items within a group.

- Groups and items may specify a `default_order` attribute which is equivalent to `order` but won't override the 
  existing value if there is one.
  
Example: You've decided yellow belts are too slow, so your mod rips them out of the game and adds Ludicrous Transport 
Belts instead.
 
A patch accomplishing this will be look like this:

```lua
-- Note that this outer table is always required!
categories = {
    belts = {
        groups = {
            -- Deletes the 'normal' group
            -- normal = false,
                
            -- Addds our new group
            ludicrous = {
                order = 400,
                items = {
                    ['ludicrous-belt'] =             { order = 100, type = 'belt' },
                    ['ludicrous-underground-belt'] = { order = 200, type = 'underground' },
                    ['ludicrous-splitter'] =         { order = 300, type = 'splitter' },
                }
            }
        }
    }
}
```

Note that it is harmless to have items in the mapper that don't have prototypes -- though it will trigger a (probably
unnoticeable) performance hit when cycling items.

#### `remote.call("QuickItemSwap", "dump_inventory", inventory)`

Developer function.  May disappear at any time.

#### `remote.call("QuickItemSwap", "dump_quickbar", player)`

Developer function.  May disappear at any time.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.3.4
Date: 06. 14. 2020

  Changes:
    - Fix crash for 'Unknown consuming type' when loading mod

---------------------------------------------------------------------------------------------------
Version: 0.3.3
Date: 01. 31. 2020

  Changes:
    - Update for Factorio 0.18

---------------------------------------------------------------------------------------------------
Version: 0.3.2
Date: 14. 03. 2019

  Changes:
    - QuickItemSwap can now cycle between item ghosts if you lack the actual item.
      - By default, this excludes items that have not been researched.  You can globally disable this in Mod Settings, or you can whitelist specific items.
    - You can now whitelist or blacklist items (default keybind: Alt+Y while holding an item)
      - When cycling between related items, blacklisted items will be ignored (unless they are in your inventory)
      - When cycling between related items, whitelisted items will be allowed (even if they are not yet researched)
      - Blacklisting and whitelisting only affects cases when a ghost would be chosen or if an item would be created (in cheat mode).  Items in your inventory are always considered.
      - You can reset your entire blacklist with /qis-clear-blacklist
   - Added support for the train stops in the Train Supply Manager mod

---------------------------------------------------------------------------------------------------
Version: 0.3.1
Date: 27. 02. 2019
  Changes:
    - Update for Factorio 0.17.
    - Removed all quickbar logic.
    - Generate group information for belts/splitters/undergrounds procedurally.  This should automatically add support for modded belts, such as those from Bob's Logistics.

---------------------------------------------------------------------------------------------------
Version: 0.3.0
Date: 23. 04. 2018
  Changes:
    - Add support for blueprints, blueprint books, deconstruction planners, and similar items from mods.
      - Items of a particular type are treated as members of one group.  "Y" will cycle between different blueprint books, for instance.
      - Cycling groups will cycle between selecting a blueprint, a blueprint book, or a deconstruction planner.
      - Other items that have an item_number are also supported when cycling items, but do not support cycling groups.
    - Added support for the Artillery Wagon
    - Add support for Electric Vehicles (locomotives only)
    - Add support for FARL (locomotives only)

---------------------------------------------------------------------------------------------------
Version: 0.2.4
Date: 23. 12. 2017
  Changes:
    - Add support for the unofficial Creative Mode Fix.
    - Add Buffer Chests to the Logistic Chests group, and reordered the group to match the inventory window.

---------------------------------------------------------------------------------------------------
Version: 0.2.3
Date: 13. 12. 2017
  Changes:
    - Update for Factorio 0.16
    - Temporarily add support for Batteries Not Included, since [a bug in the current Factorio experimental](https://forums.factorio.com/viewtopic.php?f=182&t=54567&p=321491) prevents that support from working in the opposite direction.
  Modding:
    - Added more verbose error messaging if another mod submits an invalid patch.
    - Swapped the order of arguments on `apply_patch` to make it less likely for someone to inadvertently forget to specify `source`.

---------------------------------------------------------------------------------------------------
Version: 0.2.2
Date: 11. 12. 2017
  Changes:
    - Add support for items from Logistic Train Network, Smarter Trains, Vehicle Wagon, and Nixie Tubes.
    - Add speculative support for artillery wagons.

---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 05. 12. 2017
  Changes:
    - Fix default keybindings for Previous Item and Previous Group.
    - Console announcements when a configuration change is detected are now only displayed in debug mode.  They also now are prefixed with `[QuickItemSwap]` to indicate they are coming from this mod.

---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 01. 12. 2017
  Changes:
    - Added support for items from Creative Mode
    - Added support for creating items from thin air (rather than skipping over them) when your player is in cheat mode (i.e. as a result of the creative mode mode or `/c game.players[1].cheat_mode = true`).  Can be toggled in settings.  Has no effect when not in cheat mode.
  Modding:
    - Made significant internal changes to allow mod integration
    - First pass of an API to allow other mods to add recipes.

---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 30. 11. 2017
  Changes:
    - Flip logic on modules to make more sense: each tier is now a group, rather than each module type being a group.

---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 30. 11. 2017
  Changes:
    - First public release.



qis/mappings.lua
-- Related item data and utility functions.

--[[

Item mapping hierarchy:
    CATEGORY_NAME = {                           Category name is used by the API for other mod support
        {
            name = API_NAME_OF_GROUP            Group name is used by the API for other mod support
            order = SORT_VALUE                  Where this group falls in the sort order.  Leave gaps for other mod support.
            items = {                           List of items in this group
                {
                    order = SORT_VALUE,         Item's order within this group
                    name = ENTITY_NAME,         Item's entity name as provided by Factorio
                    type = CATEGORY,            Item's type.  May be nil.  If non-nil and changing to a group that has
                                                    an item of the same type, that item is selected.  Should be unique
                                                    (other than nil) within each group, behavior is undefined otherwise
                } ...
            }
        },
        {
            name = API_NAME_OF_ANOTHER_GROUP,
            ...
        }
    }
]]

local Mappings = {}
Mappings.dirty = true
Mappings.categories = {
    belts = {
        groups = {
            normal = {
                order = 100,
                items = {
                    ['transport-belt'   ] = { order = 100, type = 'belt' },
                    ['underground-belt' ] = { order = 200, type = 'underground' },
                    ['splitter'         ] = { order = 300, type = 'splitter' },
                },
            },
            fast = {
                order = 200,
                items = {
                    ['fast-transport-belt'   ] = { order = 100, type = 'belt' },
                    ['fast-underground-belt' ] = { order = 200, type = 'underground' },
                    ['fast-splitter'         ] = { order = 300, type = 'splitter' },
                },
            },
            express = {
                order = 300,
                items = {
                    ['express-transport-belt'   ] = { order = 100, type = 'belt' },
                    ['express-underground-belt' ] = { order = 200, type = 'underground' },
                    ['express-splitter'         ] = { order = 300, type = 'splitter' },
                },
            },
        }
    },
    liquids = {
        groups = {
            pipes = {
                order = 100,
                items = {
                    ['pipe'] = { order = 100 },
                    ['pipe-to-ground'] = { order = 200 },
                }
            },
            structures = {
                order = 200,
                items = {
                    ['offshore-pump'] = { order = 100 },
                    ['pump'] = { order = 200 },
                    ['storage-tank'] = { order = 300 },
                }
            },
        }
    },
    circuits = {
        groups = {
            wires = {
                order = 100,
                items = {
                    ['green-wire'] = { order = 100 },
                    ['red-wire'] = { order = 200 },
                    ['copper-cable'] = { order = 300 },
                }
            },
            combinators = {
                order = 200,
                items = {
                    ['constant-combinator'] = { order = 100 },
                    ['arithmetic-combinator'] = { order = 200 },
                    ['decider-combinator'] = { order = 300 },
                    ['small-lamp'] = { order = 1000 },
                    ['power-switch'] = { order = 1100 },
                    ['programmable-speaker'] = { order = 1200 },
                }
            },
        }
    },
    storage = {
        groups = {
            normal = {
                order = 100,
                items = {
                    ['wooden-chest'] = { order = 100 },
                    ['iron-chest'] = { order = 200 },
                    ['steel-chest'] = { order = 300 },
                }
            },
            logistics = {
                order = 200,
                items = {
                    ['logistic-chest-active-provider'] = { order = 100 },
                    ['logistic-chest-passive-provider'] = { order = 200 },
                    ['logistic-chest-requester'] = { order = 300 },
                    ['logistic-chest-storage'] = { order = 400 },
                }
            },
        }
    },
    rail = {
        groups = {
            track = {
                order = 100,
                items = {
                    ['rail'] = { order = 100 },
                    ['train-stop'] = { order = 200 },
                }
            },
            signals = {
                order = 200,
                items = {
                    ['rail-signal'] = { order = 100 },
                    ['rail-chain-signal'] = { order = 200 },
                }
            },
        }
    },
    trains = {
        groups = {
            trains = {
                order = 100,
                items = {
                    ['locomotive'] = { order = 100 },
                    ['cargo-wagon'] = { order = 200 },
                    ['fluid-wagon'] = { order = 300 },
                    ['artillery-wagon'] = { order = 400 },  -- 0.16 Soon(tm)
                }
            },
        }
    },
    inserters = {
        groups = {
            plain = {
                order = 100,
                items = {
                    ['burner-inserter'] = { order = 100 },
                    ['inserter'] = { order = 200 },
                    ['long-handed-inserter'] = { order = 300 },
                }
            },
            fast = {
                order = 200,
                items = {
                    ['fast-inserter'] = { order = 100, type = 'normal' },
                    ['filter-inserter'] = { order = 200, type = 'filter' },
                }
            },
            stack = {
                order = 300,
                items = {
                    ['stack-inserter'] = { order = 100, type = 'normal' },
                    ['stack-filter-inserter'] = { order = 200, type = 'filter' },
                }
            },
        }
    },
    power_production = {
        groups = {
            steam = {
                order = 100,
                items = {
                    ['boiler'] = { order = 100 },
                    ['steam-engine'] = { order = 200 },
                }
            },
            solar = {
                order = 200,
                items = {
                    ['solar-panel'] = { order = 100 },
                    ['accumulator'] = { order = 200 },
                }
            },
            nuclear = {
                order = 300,
                items = {
                    ['nuclear-reactor'] = { order = 100 },
                    ['heat-exchanger'] = { order = 200 },
                    ['heat-pipe'] = { order = 300 },
                }
            },
        }
    },
    power_distribution = {
        groups = {
            only_one_group = {
                order = 100,
                items = {
                    ['small-electric-pole'] = { order = 100 },
                    ['medium-electric-pole'] = { order = 200 },
                    ['big-electric-pole'] = { order = 300 },
                    ['substation'] = { order = 400 },
                }
            }
        }
    },
    miners = {
        groups = {
            only_one_group = {
                order = 100,
                items = {
                    ['burner-mining-drill'] = { order = 100 },
                    ['electric-mining-drill'] = { order = 200 },
                }
            },
        }
    },
    assemblers = {
        groups = {
            only_one_group = {
                order = 100,
                items = {
                    ['assembling-machine-1'] = { order = 100 },
                    ['assembling-machine-2'] = { order = 200 },
                    ['assembling-machine-3'] = { order = 300 },
                }
            },
        }
    },
    furnaces = {
        groups = {
            only_one_group = {
                order = 100,
                items = {
                    ['stone-furnace'] = { order = 100 },
                    ['steel-furnace'] = { order = 200 },
                    ['electric-furnace'] = { order = 300 },
                }
            },
        }
    },
    tiles = {
        groups = {
            only_one_group = {
                order = 100,
                items = {
                    ['stone-brick'] = { order = 100 },
                    ['concrete'] = { order = 200 },
                    ['hazard-concrete'] = { order = 300 },
                    ['landfill'] = { order = 400 },
                }
            },
        }
    },
    modules = {
        groups = {
            tier1 = {
                order = 100,
                items = {
                    ['speed-module'] = { order = 100, type='speed' },
                    ['effectivity-module'] = { order = 200, type='effectivity'  },
                    ['productivity-module'] = { order = 300, type='productivity'  },
                }
            },
            tier2 = {
                order = 200,
                items = {
                    ['speed-module-2'] = { order = 100, type='speed' },
                    ['effectivity-module-2'] = { order = 200, type='effectivity'  },
                    ['productivity-module-2'] = { order = 300, type='productivity'  },
                }
            },
            tier3 = {
                order = 300,
                items = {
                    ['speed-module-3'] = { order = 100, type='speed' },
                    ['effectivity-module-3'] = { order = 200, type='effectivity'  },
                    ['productivity-module-3'] = { order = 300, type='productivity'  },
                }
            },
        }
    },
}


-- Sorts a table based on the 'order' field of its subtables.  Returns a new table.
-- The subtables are also modified to also have an 'index' field specifying their location.
local function sorted(t)
    local result = {}
    for _, v in pairs(t) do
        table.insert(result, v)
    end

    table.sort(result, function(a, b) return a.order < b.order end)

    for i, v in ipairs(result) do
        v.index = i
    end

    return result
end


-- Used temporarily during Mappings:refresh to ensure every access returns a table of some form.
local DefaultCatalogValues = {
    __index = function(table, key)
        v = {}
        table[key] = v
        return v
    end
}


function Mappings:refresh(if_dirty)
    if if_dirty and not self.dirty then
        return
    end

    self.catalog = {}
    setmetatable(self.catalog, DefaultCatalogValues)

    for name, category in pairs(self.categories) do
        category.name = name    -- To aid inspection later on
        category.sorted = sorted(category.groups)

        for name, group in pairs(category.groups) do
            group.name = name
            group.id = category.name .. "/" .. group.name
            group.sorted = sorted(group.items)
            group.category = category
            group.typed = {}

            for name,item in pairs(group.items) do
                item.name = name
                item.group = group
                item.category = category
                table.insert(self.catalog[name], item)
                if item.type then
                    group.typed[item.type] = item
                end
            end
        end
    end

    setmetatable(self.catalog, nil)
    self.dirty = false
end


function Mappings:find(name, category, group)
    --[[
     - Looks for the item 'name' in the catalog.

       If it exists, returns the first matching entry for the item

       If `category` and `group` are both specified and an entry matches them, that entry is returned instead.

       Returns nil for nonexistant items.
      ]]
    self:refresh(true)

    if category and group then
        category = self.categories[category]
        if category then
            group = category.groups[group]
            if group and group.items[name] then
                return group.items[name]
            end
        end
    end

    return (self.catalog[name] and self.catalog[name][1]) or nil
end


Mappings:refresh()
return Mappings
